pkg/scheduler/objects/queue.go

```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
 	return fmt.Sprintf("{QueuePath: %s, State: %s, StateTime: %x, MaxResource: %s}",
 		sq.QueuePath, sq.stateMachine.Current(), sq.stateTime, sq.maxResource)
 }
+
+func (sq *Queue) incRunningApps() {
+	if sq == nil {
+		return
+	}
+	if sq.parent != nil {
+		sq.parent.incRunningApps()
+	}
+	sq.internalIncRunningApps()
+}
+
+func (sq *Queue) internalIncRunningApps() {
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps++
+}
```

Comment:
This can be folded into the `incRunningApps()`

Reply:
The reason I separated this logic into a new method instead of merging into incRunningApps() is: 
Based on my understanding, instruction order is not ensured in golang even with lock  (refer: https://stackoverflow.com/questions/60340382/can-i-use-lock-to-ensure-instruction-order)
If we merge into incRunningApps(), sq.Lock() could get executed before this recursion. Each parent could be locked during recursion.  
if sq.parent != nil {
		ok = sq.parent.canRun()
}

Reply:
I want to have some deeper understanding with this.

Reply:
@lixmgl I read the stackoverflow topic and that's related to goroutine scheduling. If what you're saying is correct that would mean that sequantial consistency is broken. What you're doing here is single threaded, meaning everything runs on a single goroutine. To me it would be extremely suprirising if merging `internalIncRunningApps()` into `incRunningApps()` would result in incorrect behavior.

Reply:
@pbacsko what i did is first find the parent recursively [method](https://github.com/apache/yunikorn-core/pull/429/files#diff-27632d48eb925e150a33bc92370ceaa66c31048018d11ca7a53a0b50ab7250acR1310 ) 
Then lock this parent sq only and incRunningApps. It's not single threaded. 
This way we can avoid locking all the parents during the recursion. 
I recall @wilfred-s pointed out this issue here https://github.com/apache/yunikorn-core/pull/384/files#r843616529 for combining into one method.

Reply:
Inlining `internalIncRunningApps()` will not change the behavior in any way, because the unlock will happen prior to returning from `incRunningApps()` in either case. They are functionally equivalent. However, do be sure that the lock and defer still come immediately before incrementing `runningApps`. They should not be relocated to the top of `incRunningApps()`

Reply:
@craigcondit Thanks for the info, it make sense.
I updated by merging them into one method. Please review.